app/scripts/metamask-controller.js

```diff
@@ -2657,14 +2764,14 @@ export default class MetamaskController extends EventEmitter {
           this.networkController,
         ),
       // PreferencesController
-      setSelectedAddress: preferencesController.setSelectedAddress.bind(
-        preferencesController,
-      ),
+      setSelectedAddress: (address) => {
+        this.preferencesController.setSelectedAddress(address);
+        this.accountsController.setSelectedAccount(
+          this.accountsController.getAccountByAddress(address).id,
+        );
```

Comment:
It looks like `getAccountByAddress` can return `undefined`, do we need to check for it?

``` suggestion
        const account = this.accountsController.getAccountByAddress(address);
        if (account) {
          this.accountsController.setSelectedAccount(account.id);
        }
```

or should we throw an error?

``` suggestion
        const account = this.accountsController.getAccountByAddress(address);
        if (account) {
          this.accountsController.setSelectedAccount(account.id);
        } else {
          throw new Error("Could not find account");
        }
```

Reply:
I think throwing an error is better since it keeps the same behaviour of the same as . We should move this logic to the `accounts-controller`

Reply:
Solved in [99ecc3e](https://github.com/MetaMask/metamask-extension/pull/21437/commits/99ecc3e2e787e92837f92fd506170a2a16fcce28)